src/main/java/picocli/AutoComplete.java

```diff
@@ -534,6 +534,123 @@ public static String bash(String scriptName, CommandLine commandLine) {
         return result.toString();
     }
 
+
+    public static String fish(String scriptName, CommandLine commandLine) {
+        if (scriptName == null)  { throw new NullPointerException("scriptName"); }
+        if (commandLine == null) { throw new NullPointerException("commandLine"); }
+        List<CommandDescriptor> hierarchy = createHierarchy(scriptName, commandLine);
+        //print hierarchy
+        StringBuilder result = new StringBuilder();
+        //result.append("complete --command ").append(scriptName).append(" --no-files").append("\n");
+
+        String parentFunction = "";
+        List<CommandDescriptor> currentLevel = new ArrayList<CommandDescriptor>();
+        List<String> currentLevelCommands = new ArrayList<String>();
+
+        CommandDescriptor rootDescriptor = null;
+        for (CommandDescriptor descriptor : hierarchy) {
+            if (descriptor.parentFunctionName.equals("")) {
+                rootDescriptor = descriptor;
+                continue;
+            }
+            if (!descriptor.parentFunctionName.equals(parentFunction)) {
+                processLevel(scriptName, result, currentLevel, currentLevelCommands, parentFunction, rootDescriptor);
+                rootDescriptor = null;
+
+                currentLevel.clear();
+                currentLevelCommands.clear();
+                parentFunction = descriptor.parentFunctionName;
+            }
+
+            currentLevel.add(descriptor);
+            currentLevelCommands.add(descriptor.commandName);
+        }
+        processLevel(scriptName, result, currentLevel, currentLevelCommands, parentFunction, rootDescriptor);
+
+
+        return result.toString();
+    }
+
+    private static void processLevel(String scriptName, StringBuilder result, List<CommandDescriptor> currentLevel,
+                                     List<String> currentLevelCommands, String levelName,
+                                     CommandDescriptor rootDescriptor) {
+        if (levelName.equals("")) {
+            levelName = "root";
+        }
+        result.append("\n# ").append(levelName).append(" completion\n");
+        result.append("set -l ").append(levelName).append(" ").append(String.join(" ", currentLevelCommands)).append(
```

Comment:
`String.join` is Java 8 API, and picocli is compiled with Java 5, so we cannot use that method. We will have to write our own `join` utility method.